jedi/inference/gradual/typing.py

```diff
@@ -180,7 +180,44 @@ def name(self):
 
 
 class TypingClassValueWithIndex(_TypingClassMixin, TypingValueWithIndex):
-    pass
+    def infer_type_vars(self, value_set, is_class_value=False):
+        # Circular
+        from jedi.inference.gradual.annotation import merge_pairwise_generics, merge_type_var_dicts
+
+        type_var_dict = {}
+        annotation_generics = self.get_generics()
+
+        if not annotation_generics:
+            return type_var_dict
+
+        annotation_name = self.py__name__()
+        if annotation_name == 'Type':
+            if is_class_value:
+                for element in value_set:
+                    element_name = element.py__name__()
+                    if annotation_name == element_name:
+                        merge_type_var_dicts(
+                            type_var_dict,
+                            merge_pairwise_generics(self, element),
+                        )
+
+            else:
+                return annotation_generics[0].infer_type_vars(
+                    value_set,
+                    is_class_value=True,
+                )
+
+        elif annotation_name == 'Callable':
+            if len(annotation_generics) == 2:
+                return annotation_generics[1].infer_type_vars(
+                    value_set.execute_annotation(),
+                )
+
+        elif annotation_name == 'Tuple':
+            tuple_annotation, = self.execute_annotation()
+            return tuple_annotation.infer_type_vars(value_set, is_class_value)
```

Comment:
Is there a reason why the logic of `ProxyTypingClassValue` isn't here? Just wondering.

Reply:
Not sure I understand your question here? (I can't see any logic in `ProxyTypingClassValue`? Is that the class you mean?)

Reply:
Yes, I meant the Tuple code that's in there. I guess the tuple code is called from here.

Reply:
yeah, here `tuple_annotation` is an instance of `Tuple` and we're calling the method on that class (the code for which is the following diff block).